File: Serene/Serene.Script/Membership/ChangePasswordPanel.cs

namespace Serene.Membership
{
    using jQueryApi;
    using Serenity;
    using System.Html;

    [Panel, FormKey("Membership.ChangePassword")]
    public class ChangePasswordPanel : PropertyPanel<object>
    {
        private ChangePasswordForm form;

        public ChangePasswordPanel(jQueryObject container)
            : base(container)
        {
            form = new ChangePasswordForm(this.IdPrefix);

            form.NewPassword.AddValidationRule(this.UniqueName, e =>
            {
                if (form.ConfirmPassword.Value.Length < 7)
                    return string.Format(Q.Text("Validation.MinRequiredPasswordLength"), 7);

                return null;
            });

            form.ConfirmPassword.AddValidationRule(this.UniqueName, e =>
            {
                if (form.ConfirmPassword.Value != form.NewPassword.Value)
                    return Q.Text("Validation.PasswordConfirmMismatch");

                return null;
            });

            this.ById("SubmitButton").Click((s, e) =>
            {
                e.PreventDefault();

                if (!ValidateForm())
                    return;

                var request = GetSaveEntity();
                Q.ServiceCall(new ServiceCallOptions
                {
                    Url = Q.ResolveUrl("~/Account/ChangePassword"),
                    Request = request.As<ServiceRequest>(),
                    OnSuccess = response =>
                    {
                        Q.Information(Q.Text("Forms.Membership.ChangePassword.Success"), () =>
                        {
                            Window.Location.Href = Q.ResolveUrl("~/");
                        }, new ConfirmOptions());
                    }
                });
            });
        }
    }
}

File: Serene/Serene.Web/Modules/Membership/Account/ChangePassword/ChangePasswordPanel.ts
namespace Serene.Membership {

    @Serenity.Decorators.registerClass()
    export class ChangePasswordPanel extends Serenity.PropertyPanel<ChangePasswordRequest, any> {

        protected getFormKey() { return ChangePasswordForm.formKey; }

        private form: ChangePasswordForm;

        constructor(container: JQuery) {
            super(container);

            this.form = new ChangePasswordForm(this.idPrefix);

            this.form.NewPassword.addValidationRule(this.uniqueName, e => {
                if (this.form.w('ConfirmPassword', Serenity.PasswordEditor).value.length < 7) {
                    return Q.format(Q.text('Validation.MinRequiredPasswordLength'), 7);
                }
            });

            this.form.ConfirmPassword.addValidationRule(this.uniqueName, e => {
                if (this.form.ConfirmPassword.value !== this.form.NewPassword.value) {
                    return Q.text('Validation.PasswordConfirm');
                }
            });

            this.byId('SubmitButton').click(e => {
                e.preventDefault();

                if (!this.validateForm()) {
                    return;
                }

                var request = this.getSaveEntity();
                Q.serviceCall({
                    url: Q.resolveUrl('~/Account/ChangePassword'),
                    request: request,
                    onSuccess: response => {
                        Q.information(Q.text('Forms.Membership.ChangePassword.Success'), () => {
                            window.location.href = Q.resolveUrl('~/');
                        });
                    }
                });
            });
        }

        getTemplate() {
            return `<div class="s-Panel">
    <h3 class="page-title mb-4 text-center">${Q.text("Forms.Membership.ChangePassword.FormTitle")}</h3>
    <form id="~_Form" action="">
        <div id="~_PropertyGrid"></div>
        <div class="px-field mt-4">
            <button id="~_SubmitButton" type="submit" class="btn btn-primary w-100">
                ${Q.text("Forms.Membership.ChangePassword.SubmitButton")}
            </button>
        </div>
    </form>
</div>`;
        }
    }
}